common/src/main/java/io/pravega/common/util/SimpleDeque.java

```diff
@@ -0,0 +1,315 @@
+            expand();
+        }
+    }
+
+    /**
+     * Removes one item from the head of the {@link SimpleDeque}.
+     * See {@link java.util.Deque#pollFirst()}.
+     *
+     * @return The item at the head of the {@link SimpleDeque}, or null of empty.
+     */
+    @SuppressWarnings("unchecked")
+    public T pollFirst() {
+        T e = (T) this.items[this.head];
+        if (e != null) {
+            this.items[this.head] = null;
+            this.head = increment(this.head);
+        }
+
+        return e;
+    }
+
+    /**
+     * Removes one or more items from the head of the {@link SimpleDeque}.
+     *
+     * This method is equivalent to:
+     * <pre><code>
+     *     SimpleDeque source;
+     *     Queue target;
+     *     int count = maxCount;
+     *     while(count > 0 && !source.isEmpty()) {
+     *         target.add(source.removeFirst());
+     *         count--;
+     *     }
+     * </code></pre>
+     *
+     * This method is optimized for a bulk copy and is preferred to using the code exemplified above.
+     *
+     * @param maxCount The maximum number of items to remove. If this number is larger than {@link #size()}, then
+     *                 {@link #size()} items will be removed.
+     * @return A {@link Queue} containing the removed items, in the same order as they were in the {@link SimpleDeque}.
+     */
+    public Queue<T> pollFirst(int maxCount) {
+        int count = Math.min(maxCount, size());
+        final Object[] result = new Object[count];
+        if (count == 0) {
+            return new WrapQueue<>(result);
+        }
+
+        int copyLength = Math.min(this.items.length - this.head, count);
+        System.arraycopy(this.items, this.head, result, 0, copyLength);
+        Arrays.fill(this.items, this.head, this.head + copyLength, null);
+        this.head = this.head + copyLength;
+        if (this.head == this.items.length) {
+            this.head = 0;
+        }
+
+        count -= copyLength;
+        if (count > 0) {
+            System.arraycopy(this.items, 0, result, copyLength, count);
+            Arrays.fill(this.items, 0, count, null);
+            this.head += count; // count is less than size(), so this can't wrap around no matter what.
+        }
+
+        return new WrapQueue<>(result);
+    }
+
+    /**
+     * Returns (but does not remove) the item at the head of the {@link SimpleDeque}.
+     * This method does not alter the internal state of the {@link SimpleDeque}.
+     * See {@link java.util.Deque#peekFirst()}.
+     *
+     * @return The item at the head of the {@link SimpleDeque}, or null if empty.
+     */
+    @SuppressWarnings("unchecked")
+    public T peekFirst() {
+        return (T) this.items[this.head];
+    }
+
+    /**
+     * Gets a value indicating the number of items in the {@link SimpleDeque}.
+     *
+     * @return The size of the {@link SimpleDeque}.
+     */
+    public int size() {
+        final int size = this.tail - this.head;
```

Comment:
The correctness of this function and several others are predicated on the fact that there is always at least one empty position in the array. This appears to be enforced but it probably worth explicitly stating in a comment somewhere.

Reply:
I updated the class Javadoc with implementation details.